core/ir/aarch64/codec.c

```diff
@@ -1744,6 +1852,143 @@ encode_opnd_z16(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_z(16, opnd, enc_out);
 }
 
+/* fpimm13: floating-point immediate for scalar fmov */
+
+static inline bool
+decode_opnd_fpimm13(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    /*
+     * From the Architecture Reference Manual, 8 bit immediate abcdefgh maps to
+     * floats:
+     *
+     * 3332 2222 2222 1111 1111 11
+     * 1098 7654 3210 9876 5432 1098 7654 3210
+     *  _                            abcd efgh <- 8 bit immediate mapped to
+     * abbb bbbc defg h000 0000 0000 0000 0000 <- 32 bit float
+     *
+     *   abcd efgh  Masks
+     * 0x1    0     a
+     * 0x4    0     b
+     * 0x2    0     c
+     * 0x1    F     defgh
+     */
+    if (extract_uint(enc, 22, 1) == 0) { /* 32 bits */
+        union {
+            float f;
+            uint32_t i;
+        } fpv;
+
+        uint32_t imm = extract_uint(enc, 13, 8);
+
+        uint32_t a = imm & 0x80;
+        uint32_t b = imm & 0x40;
+        uint32_t not_b = ((b == 0) ? 1 : 0);
+        uint32_t bbbbb = ((b == 0) ? 0 : 0x1f);
+        uint32_t c = imm & 0x20;
+        uint32_t defgh = imm & 0x1f;
+
+        uint32_t imm32 =
+            (a << 24) | (not_b << 30) | (bbbbb << 25) | (c << 19) | (defgh << 19);
+
+        fpv.i = imm32;
+        *opnd = opnd_create_immed_float(fpv.f);
+    } else { /* 64 bits */
+        /* 6666 5555 5555 5544 44444444 33333333 33322222 22221111 111111
+         * 3210 9876 5432 1098 76543210 98765432 10987654 32109876 54321098 76543210
+         *  _                                                               abcdefgh
+         * abbb bbbb bbcd efgh 00000000 00000000 00000000 00000000 00000000 00000000
+         */
+        union {
+            double f;
+            uint64_t i;
+        } fpv;
+
+        uint64_t imm = extract_uint(enc, 13, 8);
+
+        uint64_t a = imm & 0x80;
+        uint64_t b = imm & 0x40;
+        uint64_t not_b = ((b == 0) ? 1 : 0);
+        uint64_t bbbbbbbb = ((b == 0) ? 0 : 0xff);
+        uint64_t c = imm & 0x20;
+        uint64_t defgh = imm & 0x1f;
+
+        uint64_t imm64 =
+            (a << 56) | (not_b << 62) | (bbbbbbbb << 55) | (c << 48) | (defgh << 48);
+
+        fpv.i = imm64;
+        *opnd = opnd_create_immed_float(fpv.f);
```

Comment:
Hmm, opnd_create_immed_float only takes a float, so this will lose precision from the double.  It seems we need to add a double value to opnd_t.  For 64-bit, replacing the existing float with double should result in no size change to opnd_t: for 32-bit I guess it will, but if no ISA has double immediates for 32-bit we could have an ifdef for X64?

Reply:
Done.